v6 beta branch

yoshuawuyts opened the pull request:
in-progress target branch we can use for breaking behavior. The prototype branch is basically a full rewrite; this allows us to put out some betas and iterate on the minor changes we want for the next version.

Basically I'm thinking we keep the API stable, but yeah there's a few nits that are technically breaking. So we should iron those out in a semver major. Yay for boring changes :sparkles:

### related issues
- [x] call DOMcontentloaded on next frame https://github.com/yoshuawuyts/choo/pull/465
- [x] use nanotiming https://github.com/yoshuawuyts/choo/pull/476
- [x] become prototype based https://github.com/yoshuawuyts/choo/pull/480
- [x] add tracing to nanobus https://github.com/yoshuawuyts/nanobus/pull/10
- [x] revert special casing of root onload handlers https://github.com/yoshuawuyts/nanomorph/pull/65
- [x] fix select tags https://github.com/yoshuawuyts/nanomorph/pull/60
- [x] remove onload https://github.com/shama/bel/pull/69
- [x] should clear performance timers https://github.com/yoshuawuyts/choo/issues/461
- [x] remove nanomount https://github.com/yoshuawuyts/choo/issues/481
- [x]  remove nanohistory (too small a module)
- [x] fix whitespace in `bel`
- [ ] fix whitespace in `yo-yoify` https://github.com/shama/yo-yoify/issues/49
- [ ] fix whitespace in `pelo` https://github.com/shuhei/pelo/issues/7
- [ ] upgrading bankai to new `yo-yoify`
- [ ] upgrading `bel` to new `pelo`

yoshuawuyts commented:
Timing API has been upgraded, now shows separate timings for creating DOM trees + morphing the trees.

<img width="680" alt="screen shot 2017-05-09 at 13 35 01" src="https://cloud.githubusercontent.com/assets/2467194/25848829/61c6cca4-34bc-11e7-9be6-7808645312ad.png">

I'm not sure about the event names tho; right now they are:
```txt
choo.render:view.route   # for choo's internal stuff
choo.emitter:foo:bar     # for an event a user registers
```
What I'm wondering is if we should perhaps upgrade [nanobus](https://github.com/yoshuawuyts/nanobus/) to delimit internal events with a `"/"` so it'd become:

```txt
choo:render/view:route   # for choo's internal stuff
choo:emitter/foo:bar     # for an event a user registers
```

Thoughts? cc/ @juliangruber @timwis

yoshuawuyts commented:
note to self: move the new choo `performance.measure()` to `requestIdleCallback` and use https://github.com/yoshuawuyts/nanotiming/pull/1

juliangruber commented:
thoughts:

- `[choo] render: view.route`
- `choo.render(view.route)`
- `[choo] render(view.route)`

moszeed commented:
i like the following proposals by @juliangruber 
- choo.render(view.route)
- [choo] render(view.route)

yoshuawuyts commented:
@juliangruber gonna try some of those out; like the `choo.render(view.route)` idea for sure! - should become part of https://github.com/yoshuawuyts/nanotiming/pull/3 I reckon

yoshuawuyts commented:
Also thinking of adding a `navigate` event which is fired whenever path changes, and returns the current name of the route. This could be used in the logger as a `debug` style event - and would be useful for APM providers so they can hook into the actual name of the route.

tornqvist commented:
Is there any reason for `createLocation` to not be exposed on the prototype? It would kind of solve #440 as you could roll your own method that ignores hashes and then simply set `opts.href` to false and implement nanohref in a similar way.

yoshuawuyts commented:
@tornqvist I think what you're proposing makes sense (:

yoshuawuyts commented:
Last issues remaining:
- https://github.com/shama/yo-yoify/issues/49
- https://github.com/shuhei/pelo/issues/7
- upgrading bankai to new `yo-yoify`
- upgrading `bel` to new `pelo`

juliangruber commented on CHANGELOG.md:
```diff
@@ -1,3 +1,34 @@
+you ever need to.
+
+We've also tweaked the event system a little. All built-in event names are now
+available as constants under `choo.events`. And we've introduced three new
```
three or two?

yoshuawuyts replied:
Oops, good point - it's two brand new ones, and one that was introduced a
couple of weeks ago and folks might not be aware of. Will fix! :D

yoshuawuyts replied:
Fixed!

yoshuawuyts commented:
Oops, good point - it's two brand new ones, and one that was introduced a
couple of weeks ago and folks might not be aware of. Will fix! :D

On Mon, Jun 26, 2017 at 8:44 PM Julian Gruber <notifications@github.com>
wrote:

> *@juliangruber* commented on this pull request.
> ------------------------------
>
> In CHANGELOG.md
> <https://github.com/yoshuawuyts/choo/pull/489#discussion_r124101940>:
>
> > +things will work great.
> +
> +Choo v6 includes the upgrades to `bel@5` and `nanomorph@5`. This means up to
> +20x faster server rendering, and up to 10x improved browser rendering speeds.
> +We've also fixed DOM node caching, and introduced sibling node reordering.
> +
> +This release also includes an overhauled timing API. We're now capturing more
> +events, and are leveraging this in tools like `choo-log` to provide consistent
> +feedback on performance.
> +
> +Choo is now also completely prototype based. It not only means faster execution
> +times, and more maintainable codebase - but you can also override methods if
> +you ever need to.
> +
> +We've also tweaked the event system a little. All built-in event names are now
> +available as constants under `choo.events`. And we've introduced three new
>
> three or two?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/yoshuawuyts/choo/pull/489#pullrequestreview-46362752>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACWlekBJaudKFAyv35xta0Lga699lZ8Gks5sIAoUgaJpZM4NUT13>
> .
>

yoshuawuyts commented:
📦 v6.0.0-0 (release candidate)

Please share any feedback here https://github.com/yoshuawuyts/choo/issues/519. Thanks! :tada:
